### PR TITLE
Add APNG Encode Support

### PR DESCRIPTION
**Pull request by rrbutani**
Missing support for a couple of features, but seems to work.

This probably isn't ready to merge as is, but I thought I'd make the PR anyways to get some feedback.

### CODE REVIEW

**Comment by abonander** [comment]
@rrbutani can you rebase against latest master? Looks like there's some merge conflicts.

@bvssvni @nox I've run into a use-case for this, I'd like to see some movement on it. I'll be testing this fork myself to see how it works once it's been rebased.

**Comment by bvssvni** [comment]
@abonander Nice if you can get a look.

**Comment by abonander** [comment]
@rrbutani `FrameControl` should be on a per-frame basis since each frame can have configurable parameters. In fact, I would almost suggest an entirely different `Encoder` struct that's constructed directly with an `AnimationControl` instance.

File: src/encoder.rs [hunk 1]
```diff
@@ -136,7 +191,110 @@ impl<W: Write> Writer<W> {
...
+            Info { animation_control: Some(AnimationControl { num_frames: 0, ..}) , frame_control: Some(_), ..} => {
+                Err(EncodingError::Format("exceeded number of frames specified".into()))
+            },
+            Info { animation_control: Some(anim_ctl), frame_control: Some(frame_control), ..} => {
+                match frame_control.sequence_number {
+                    0 => {
+                        let ret = if self.separate_default_image { // If we've already written the default image we can write frames the normal way
+                            // fcTL + fdAT
+
+                            self.write_fcTL().and(self.write_fdAT(data))
```
Comment (abonander) [inline src/encoder.rs]: This kind of chaining could lead to data corruption as the `write_fdAT()` operation will be executed regardless of whether or not `write_fcTL()` succeeds. These should instead be wrapped with `try!()` like all the other fallible operations (although the `?` operator has been stable for a long time now, `try!()` would be consistent with the rest of the crate).

File: src/encoder.rs [hunk 2]
```diff
@@ -136,7 +191,110 @@ impl<W: Write> Writer<W> {
...
+        (&mut data[22..]).write_be(frame_ctl.delay_den)?;
+        data[24] = frame_ctl.dispose_op as u8;
+        data[25] = frame_ctl.blend_op as u8;
+
+        self.write_chunk(chunk::fcTL, &data)
+    }
+
+    #[allow(non_snake_case)]
+    fn write_fdAT(&mut self, data: &[u8]) -> Result<()> {
+        // println!("Writing fdAT:{:?}", self.info.frame_control.unwrap().sequence_number+1);
```
Comment (abonander) [inline src/encoder.rs]: These printlns should be removed once the PR is ready to be merged.

**Comment by rrbutani** [comment]
Thanks for the feedback @abonander; I'll rebase and make the suggested changes later today.

**Comment by fintelia** [comment]
Is this ready to be merged? (Other than the couple conflicts)

**Comment by HeroicKatora** [comment]
There seem to be two potentially independent changes here:

* Interpreting `DisposeOp` and `BlendOp` and erroring accordingly during decoding.
* Support writing animated data stream and encode `fcTL` block.

As far as I can see only the former is SemVer relevant while the latter can be done in a point-release after the first has landed. I'd like to release a new version of `image-png` and `image` relatively soon so it would be nice to either get this merged but only having the public representation changes ready is also fine.

**Comment by Rimpampa** [comment]
I opened a pull request on the @rrbutani fork: https://github.com/rrbutani/image-png/pull/1

I made it work but some features are missing, like the option to set a different framerate for each frame or dividing each frame imago into more fcTL chunks (like with the `StreamWriter`)

I'm willing to continue contribuiting on this but I'd like to hear some opinions on how to make it better, as right now it doesn't feel right.

**Comment by HeroicKatora** [comment]
@Rimpampa I think you should make your own PR so we don't need to wait on the fork's author to respond (it's been quite a while). Then also rebase those commits to resolve the merge conflict. Let me know if you'd like any help with that, I've done it once before on this one :)

**Comment by HeroicKatora** [comment]
Closed/Moved to #255

**Pull request closed without merging**